### PR TITLE
Reagent materials again

### PR DESCRIPTION
**Pull request by KayeArray**
:cl: banthebantz
add: A new room has been added to Engi maint containing various machines, this room is engi only
add: The first of note are two reagent sheet machines which can turn solid reagents into reagent ingots
add: The second is the big forge which can make various custom weapons and bullets out of ingots
add: These weapons can apply reagents on hit and have varying stats based on what reagent you use
add: The third machine of note is a reagent material trait manipulator
add: Some reagents have special traits, these can be seen in the aforementioned machine
add: This machine can add additional traits to forged weapons, fill it with a beaker and add a weapon
add: Reagent ingots can also be used to make reagent floors, walls and doors
add: These also have custom effects, when touched, shot at and heated, please be careful ;)
tweak: Solid chems no longer react with turfs when splashed
/:cl:

It should work again now and it has sprites from Fury

old: https://github.com/HippieStation/HippieStation/pull/4629

### CODE REVIEW

**Comment by ghost** [comment]
absolute madman overhauls materials

**Comment by ghost** [comment]
Local Spaceman Somehow Sneaks Forge On-Station Again
+1

**Comment by ghost** [comment]
Had +2 in the old PR

**Comment by bardofrage** [comment]
+1

File: code/__DEFINES/~hippie_defines/forge.dm [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+					F.sharpness = IS_SHARP_ACCURATE
+			if(FORGED_MELEE_TWOHANDED)
+				var/obj/item/twohanded/forged/F = I
+				F.armour_penetration += 10
+				F.stabby += TRANSFER_SHARP_BONUS
+				if(F.sharpness == IS_SHARP)
+					F.sharpness = IS_SHARP_ACCURATE
+			if(FORGED_BULLET_CASING)
+				var/obj/item/projectile/bullet/forged/F = I
+				F.dismemberment += 20
```
Comment (JohnGinnane) [inline code/__DEFINES/~hippie_defines/forge.dm:102]: Shouldn't this also add armour penetration?

File: code/__DEFINES/~hippie_defines/forge.dm [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+				START_PROCESSING(SSobj, F)
+			if(FORGED_BULLET_CASING)
+				var/obj/item/projectile/bullet/forged/F = I
+				F.radioactive = TRUE
+
+
+/datum/special_trait/ultradense
+	name = "Ultradense"
+	desc = "Found in exceptionally dense materials, allows bullets to penetrate and adds knockback to all weapons"
+	effectiveness = 100
```
Comment (JohnGinnane) [inline code/__DEFINES/~hippie_defines/forge.dm]: Redundant homey

File: code/__DEFINES/~hippie_defines/forge.dm [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+	desc = "Stats randomly change when applied"
+	effectiveness = 100//so it's truly random and you can't get stuck with an OP weapon for an extended period of time
+
+
+/datum/special_trait/unstable/on_hit(atom/target, mob/user, obj/item/I, type)
+	if(type && I)
+		switch(type)
+			if(FORGED_MELEE_SINGLEHANDED)
+				var/obj/item/forged/F = I
+				F.force = rand(1, 100)
```
Comment (JohnGinnane) [inline code/__DEFINES/~hippie_defines/forge.dm]: \>RNG
ree

**Comment by ghost** [comment]
That is +3, I'll merge this once you answer the questions from @JohnGinnane

**Comment by JohnGinnane** [comment]
I've only reviewed some of it by the way, I'll be doing more later

File: code/__DEFINES/~hippie_defines/forge.dm [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+			M.IgniteMob()
+
+	if(user && iscarbon(user))
+		var/mob/living/carbon/C = user
+		if(C.gloves)
+			return
+		var/hit_zone = (C.held_index_to_dir(C.active_hand_index) == "l" ? "l_":"r_") + "arm"
+		var/obj/item/bodypart/affecting = C.get_bodypart(hit_zone)
+		if(affecting)
+			if(affecting.receive_damage(0, 5))
```
Comment (JohnGinnane) [inline code/__DEFINES/~hippie_defines/forge.dm:257]: Should we receive burn damage? Can wearing black gloves be a method to stop this?

File: code/__DEFINES/~hippie_defines/forge.dm [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+#define FORGED_MELEE_TWOHANDED "twohanded"
+#define FORGED_BULLET_CASING "bullet"
+#define FORGED_INGOT = "ingot"
+
+//SPECIAL TRAIT DATUMS
+
+/datum/special_trait
+	var/name//these are for analysis in game so people can actually see what the traits are
+	var/desc//
+	var/effectiveness = 100//probability of the trait being activated if it's a constant or on hit effect
```
Comment (JohnGinnane) [inline code/__DEFINES/~hippie_defines/forge.dm:46]: I'm not seeing where this is actually used. Like fire has it set to 10, so is it 10% chance to ignite?

File: code/__DEFINES/~hippie_defines/forge.dm [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+			return
+		if(ishuman(C))
+			var/mob/living/carbon/human/H = C
+			if(H.dna && H.dna.species)
+				if(PIERCEIMMUNE in H.dna.species.species_traits)
+					return
+		var/hit_zone = (C.held_index_to_dir(C.active_hand_index) == "l" ? "l_":"r_") + "arm"
+		var/obj/item/bodypart/affecting = C.get_bodypart(hit_zone)
+		if(affecting)
+			if(affecting.receive_damage(0, 5))
```
Comment (JohnGinnane) [inline code/__DEFINES/~hippie_defines/forge.dm:312]: Likewise, can some gloves (maybe yellow or latex) prevent this?

File: code/__DEFINES/~hippie_defines/forge.dm [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+
+/datum/special_trait/fluidic/on_hit(atom/target, mob/user, obj/item/I, type)
+	if(I && type)
+		switch(type)
+			if(FORGED_MELEE_SINGLEHANDED)
+				if(user)
+					var/obj/item/forged/F = I
+					for(var/atom/A in view(2, user))
+						if(isliving(A))
+							F.reagent_type.reaction_mob(A, TOUCH, 2)
```
Comment (JohnGinnane) [inline code/__DEFINES/~hippie_defines/forge.dm]: Does this use up any of the reagent?

File: code/__DEFINES/~hippie_defines/forge.dm [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+						else if(isturf(A))
+							F.reagent_type.reaction_turf(A, TOUCH, 2)
+						else if(isobj(A))
+							F.reagent_type.reaction_obj(A, TOUCH, 2)
+
+			if(FORGED_MELEE_TWOHANDED)
+				if(user)
+					var/obj/item/twohanded/forged/F = I
+					for(var/atom/A in view(2, user))
+						if(isliving(A))
```
Comment (JohnGinnane) [inline code/__DEFINES/~hippie_defines/forge.dm]: Might be an idea to set up a new proc for this code as it's the same in all three cases. Search for people in 2 tile radius and apply reagent them, etc.

File: code/__DEFINES/~hippie_defines/forge.dm (hunk 4 above)
Comment (KayeArray) [reply code/__DEFINES/~hippie_defines/forge.dm:257]: if(C.gloves)
 +			return

File: code/__DEFINES/~hippie_defines/forge.dm (hunk 6 above)
Comment (KayeArray) [reply code/__DEFINES/~hippie_defines/forge.dm:312]: if(C.gloves)
 +			return

File: code/__DEFINES/~hippie_defines/forge.dm (hunk 7 above)
Comment (KayeArray) [reply code/__DEFINES/~hippie_defines/forge.dm]: no, reagent materials don't actually contain reagents they just statically apply them when activated

File: code/__DEFINES/~hippie_defines/forge.dm (hunk 1 above)
Comment (KayeArray) [reply code/__DEFINES/~hippie_defines/forge.dm:102]: nah i'd rather not

File: code/__DEFINES/~hippie_defines/forge.dm (hunk 5 above)
Comment (KayeArray) [reply code/__DEFINES/~hippie_defines/forge.dm:46]: yes

File: code/__DEFINES/~hippie_defines/forge.dm (hunk 6 above)
Comment (KayeArray) [reply code/__DEFINES/~hippie_defines/forge.dm:312]: if(C.gloves)
 +			return

File: code/__DEFINES/~hippie_defines/forge.dm (hunk 4 above)
Comment (JohnGinnane) [reply code/__DEFINES/~hippie_defines/forge.dm:257]: oh shit my 🐝

File: code/__DEFINES/~hippie_defines/forge.dm (hunk 3 above)
Comment (KayeArray) [reply code/__DEFINES/~hippie_defines/forge.dm]: i'll nerf the range a bit but that's intended

File: hippiestation/code/game/objects/items/forged_weapons.dm [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+		var/armour_block = C.getarmor(affecting, "melee") * 0.01
+		if(!armour_block)
+			armour_block = 1
+		C.reagents.add_reagent(reagent_type.id, max(0, (0.2 * stabby) * max(1, armour_block - armour_penetration)))
+		if(stabby < 1 && stabby > 0)
+			reagent_type.reaction_mob(C, TOUCH, max(0, 1 / stabby))
+	if(proximity_flag && reagent_type)
+		for(var/I in special_traits)
+			var/datum/special_trait/A = I
+			if(prob(A.effectiveness))
```
Comment (JohnGinnane) [inline hippiestation/code/game/objects/items/forged_weapons.dm:55]: GitHub didn't load all the files for me initially, so I see this now

**Comment by xTrainx** [comment]
i dont see any lean special trait

File: hippiestation/code/modules/reagents/chemistry/reagents/density.dm [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+	special_traits = list(SPECIAL_TRAIT_METALLIC, SPECIAL_TRAIT_RADIOACTIVE)
+
+/datum/reagent/radium
+	special_traits = list(SPECIAL_TRAIT_RADIOACTIVE)
+
+/datum/reagent/toxin/radgoop
+	special_traits  = list(SPECIAL_TRAIT_RADIOACTIVE)
+
+/datum/reagent/arclumin
+	special_traits  = list(SPECIAL_TRAIT_RADIOACTIVE, SPECIAL_TRAIT_MAGNETIC)
```
Comment (JohnGinnane) [inline hippiestation/code/modules/reagents/chemistry/reagents/density.dm:81]: Down below you set this to something else...

File: hippiestation/code/modules/reagents/chemistry/reagents/density.dm [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+//MISC
+
+/datum/reagent/oxygen
+	special_traits = list(SPECIAL_TRAIT_MAGNETIC)
+
+/datum/reagent/water
+	special_traits = list(SPECIAL_TRAIT_REFLECTIVE)
+
+/datum/reagent/arclumin
+	special_traits = list(SPECIAL_TRAIT_UNSTABLE)
```
Comment (JohnGinnane) [inline hippiestation/code/modules/reagents/chemistry/reagents/density.dm:148]: ...here

File: hippiestation/code/game/objects/items/forged_weapons.dm [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+/obj/item/ammo_casing/forged/proc/assign_properties()//placeholder proc to prevent runtimes, this SHOULD be the only exception to the rule
+	if(reagent_type)
+		name = "([reagent_type.name]-[caliber] bullet casing)"
+	return
+
+
+/obj/item/ammo_casing/forged/attack_self(mob/user)
+	..()
+	if(!caliber)
+		caliber = input("Shape the bullet to which caliber? (You may only do this once!)", "Transform", caliber) as null|anything in calibers
```
Comment (JohnGinnane) [inline hippiestation/code/game/objects/items/forged_weapons.dm:305]: Shouldn't you check `if (caliber)`? If user hits cancel will it still do the below code?

**Comment by JohnGinnane** [comment]
looks mostly good, i'd say give it a spin and see what happens

also conflict ree

**Comment by KayeArray** [comment]
map conflicts make me want to die

File: code/__DEFINES/~hippie_defines/forge.dm (hunk 4 above)
Comment (ghost) [reply code/__DEFINES/~hippie_defines/forge.dm:257]: f-f-fingerless gloves...

**Comment by ghost** [comment]
@banthebantz fix the conflicts and we'll testmerge

**Comment by GuyonBroadway** [comment]
figs gonfligds :DDDDDDDDDDD

No really fix dese conflicts pls.

**Comment by JohnGinnane** [comment]
@Chnkr @banthebantz 

If conflicts are sorted can we testmerge pls

**Comment by GuyonBroadway** [comment]
this is fukken STALE!!!

**Comment by Pyko1** [comment]
too stale bambipants!

**Comment by ghost** [comment]
reeee

**Comment by praisenarsie** [comment]
@banthebantz i assume you gave up on this (for now)

**Comment by praisenarsie** [comment]
also i'll keep the branch but if nothing happens in like a week or two then i'll kill it

**Comment by GuyonBroadway** [comment]
leave the branch imma steal it and fix it myself.

**Comment by praisenarsie** [comment]
👐 my hands are off that twig

**Comment by praisenarsie** [comment]
haha the branch is mine now ✋ 🌴 

👉 📕

**Pull request closed without merging**